Port tasks to new rcx-version

newhinton opened the pull request:
This updates the tasks-branch from rclone to rcx and the changes made since the fork

Closes #146, #38, #61 and i guess it closes #128 aswell

Todo: 

- [x] Migrate Export/Import to JSON
- [x] -> when making an export, it should export tasks, triggers and the rclone config as a zip file
- [x] Switch Export filestorage to non-deprecated one (https://github.com/x0b/rcx/pull/37#discussion_r406750606)
- [x] Use proper logging (https://github.com/x0b/rcx/pull/37#discussion_r406751537)
- [x] Fix naming in DatabaseHandler
- [x] (Go over all of this: https://github.com/x0b/rcx/commit/2a19150d984339d470afc4dda488934db254bf0d)
- [x] Fix Crash in Triggerfragment when no tasks are defined
- [ ] Discuss naming of things (tasks/trigger)
- [ ] Fix theming not applied in new Activities (I dont quite get why it's not working, it should)
- [x] Add SharedPreferences to backup.zip
- [x] After restoring backups, the app crashes because restoring prefs introduces an error to taskdecription (#157)
- [x] when no tasks are defined, the button linking to tasks should not only open the tasklist but the taskcreation activity aswell
- [x] also, maybe add a cool message that there are no tasks?
- [x] remove logging that was left somewhere
- [x] task list icons are not properly stored/used
- [x] fix back button behaviour on fragments
- [x] Investigate why disabled-state is beeing ignored by the trigger
- [x] Add task-stacking, so that tasks scheduled at the same time are not lost


In the future:
I want to investigate how exactly rcx behaves when multiple tasks are run simultaneous. The resulting information can be used to extend the trigger mechanism, so that it would be possible to run a task after a different task was run
- [ ] extend triggermechanism (maybe)
- [ ] Folderpicker for remotes (In the future. Maybe. Maybe not.)
- [ ] Status Indicator for task (Last run succeded, failed or is currently running)
- [ ] Add special task to export backup to remote so that rcx can backup itself
- [ ] Check regularly for storage access, because tasks "fail silently" with successmessages when it cannot access internal storage

Those ideas will very likely not be in this PR

newhinton committed:
Port tasks to new rcx-version

newhinton committed:
Re-expose TaskStartService

newhinton committed:
Migrate from TaskDialog to TaskActivity

newhinton committed:
Added Folderpicker to TaskEdit for local

newhinton committed:
fix 'Only Wifi Upload' not working

newhinton commented:
This also fixes a small bug which generally blocks uploading when it only should block it when not on wifi

LukeCode commented:
Works very well! Thanks for your work!

iamtalhaasghar reviewed (commented):
@newhinton There are conflicts which need to be resolved manually. 
Is this feature similar to rclone-browser's tasks?

newhinton commented:
@iamtalhaasghar I need to update my branch to the latest anyway, so i'll do that. 

Can you give me a link to the rclone-browser you mean?

newhinton committed:
Merge remote-tracking branch 'x0b/master' into feature_update_upstream

# Conflicts:
#	app/build.gradle
#	app/src/main/AndroidManifest.xml
#	app/src/main/res/values/strings.xml
#	gradle/wrapper/gradle-wrapper.properties

iamtalhaasghar commented:
https://kapitainsky.github.io/RcloneBrowser/

newhinton committed:
Implement userinterface to trigger tasks

newhinton commented:
Yes, i believe they are quite similar. My PR allows you to define a folder and a target, which can then be triggered by an external app.

Im currently working on a trigger-system, so that tasks can directly be scheduled via rcx, i expect this to be working shortly

newhinton committed:
Implement TriggerService to take care of starting tasks

newhinton committed:
update layout to make it look good

newhinton committed:
set sensible defaults

newhinton committed:
Finalize technical implementation of TriggerService

newhinton commented:
This now contains a trigger mechanism in addition to tasks

newhinton commented:
Todo: 

Migrate Export/Import to JSON
-> when making an export, it should export tasks, triggers and the rclone config as a zip file
Switch Export filestorage to non-deprecated one (https://github.com/x0b/rcx/pull/37#discussion_r406750606)
Use proper logging (https://github.com/x0b/rcx/pull/37#discussion_r406751537)

Fix naming in DatabaseHandler
(Go over all of this: https://github.com/x0b/rcx/commit/2a19150d984339d470afc4dda488934db254bf0d)

newhinton committed:
Overhaul DatabaseHandler and TriggerService

newhinton committed:
Overhaul Task

newhinton committed:
Overhaul Task UI

newhinton committed:
Overhaul TaskStartService

newhinton committed:
Moved Import/export to json, but removed it from ui since writing to file does not work yet

newhinton committed:
Verify existence of rclone library on each call, so that the app never crashes if the rclone binary is broken/unavailable and throw fitting error

newhinton committed:
update README.md for more development information

newhinton committed:
update TriggerItem visuals and add quick-disable to iconbutton

newhinton commented:
I capsuled the rclone processcall, because when it was not supplied by a build, the app just crashes. We should never crash, and this should trigger a more or less appropriate errormessage. (https://github.com/x0b/rcx/pull/84/commits/8d4b1228fb3063d0839a3e4c41da127a2f7b185f)
This was a bit offtopic from the initial pr, but to be honest, this has gotten out of hand anyway :D

newhinton committed:
handle databaseerrors gracefully instead of crashing

newhinton committed:
Show errorscreen when no tasks are defined to prevent userconfusion

newhinton committed:
remove Import/Export from Trigger/Task-Fragments

newhinton committed:
use zipfile for full config export

newhinton committed:
Switch import to zipfile-import

newhinton committed:
undo unintentionally pushed gradle changes

newhinton committed:
fix wrong textcolor in darkmode in trigger/tasklists

newhinton committed:
add translations to tasks_item

newhinton committed:
add de-translations

newhinton committed:
add back button to trigger/task-activity

newhinton committed:
add shared prefs to backupfile

newhinton committed:
Fix all sorts of unused imports, lamdas and codestyling issues

newhinton committed:
only show .zip files on import

newhinton committed:
overhaul fragment_tasks_item to be more consistent

newhinton committed:
implement basic task-logging (WIP)

newhinton committed:
updated log_item
fixed handling of filewriting in SyncLog
Added some places where logging should occur

newhinton committed:
fix width of content in log_item

newhinton commented:
@x0b I think this particular pr is more or less done. It needs a review (i already have solved a lot of the initial issues) and then it can be merged! However, since this is a rather large PR, please feel free to make suggestions how i can make it easier to review :D

newhinton committed:
switch rclone command building from array to arraylist

newhinton committed:
switch logging to json and remove clunky notificationbuilding

newhinton committed:
Open SyncLog when notification was pressed

newhinton committed:
remove logging and update task-menuicon

newhinton committed:
remove logging

newhinton committed:
start TaskActivity from TriggerFragment
show appropriate messages when no tasks are available in taskfragment

newhinton committed:
properly show task-icon from remote

newhinton committed:
disable restoration of colors from backup until they do not rely on explicit resource id's anymore. Also removed support for TaskDescription-theming in devices pre 23 (Android 6)

newhinton committed:
added better screens to Task/TriggerFragments and refreshed the views in edgecases so that always the proper view is shown

newhinton committed:
handle onBackPressed when in Fragment

x0b commented:
That's a huge set of changes 😮. It'll take a while to review, It is currently finals season at my university 🙈. 

It seems your PR contains fixes for unrelated or only slightly related issues - can you factor them out into their own PRs? That way, I can fast-track those fixes. To be honest, this PR is currently to large to merge ([a](https://smallbusinessprogramming.com/optimal-pull-request-size/) [few](https://tekin.co.uk/2020/05/proof-your-thousand-line-pull-requests-create-more-bugs) [reasons](https://softwareengineering.stackexchange.com/questions/235046/dealing-with-a-large-pull-request) [why](https://engineering.zalando.com/posts/2017/10/a-plea-for-small-pull-requests.html) [that's](https://essenceofcode.com/2019/10/29/the-art-of-small-pull-requests/) [bad](https://opensource.com/article/18/6/anatomy-perfect-pull-request)).

If you have fixed a known issue (e.g. https://github.com/x0b/rcx/pull/84/commits/e559beaef8b28c7ad6bc9032c226d6a5b8d3c125 seems like a fix for #127), try to reference relevant issues in your commit message ([example](https://github.com/x0b/rcx/commit/222985626ced4202f9032e1d92bbe7a57ef86f2c), the main rclone repo [contains more helpful guidance on that topic](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages)).

newhinton commented:
@x0b yes i know, the main featureset (tasks/triggers) are most of the changes, but there are changes that are very close to it (eg. the switch from multiple impor/export actionts to a unified zip containing everything). And the next major thing (logging) is already in a new pr, and i would really like to overhaul darkmode which i will tackle sometime after this pr has been merged.

Gladly, a huge chunk was already reviewed in the older pr's albeit they have undergone many changes (mainly stuff from the older review and refactoring and streamlining) 

Also: do take your time! I know this has gotten out of hand, but since i needed to implement this for myself i did it here. If it is helpful, we could make a cut at c824f38 but this would basically mean that you'd have to review many things twice since i continuously improved everything as far as i could see.  (that pr basically seperates the task and trigger logic)

newhinton committed:
switch logging to json and remove clunky notificationbuilding

Signed-off-by: Felix Nüsse <felix.nuesse@t-online.de>

newhinton committed:
add remote folderpicker workaround via SAF

Signed-off-by: Felix Nüsse <felix.nuesse@t-online.de>

newhinton commented:
I have added a workaround for the missing remote-folderpicker. It relies on SAF beeing enabled and the system filepicker, and throws errortoasts if the user doesnt select a valid target. 
I am really not sure if we should keep it, however it 'works'

newhinton committed:
implement human readable conversion

Signed-off-by: Felix Nüsse <felix.nuesse@t-online.de>

newhinton committed:
switch humanReadableBytes to builtin and make json-parsing more resilient

Signed-off-by: Felix Nüsse <felix.nuesse@t-online.de>

newhinton committed:
add logging when json-reading failed

Signed-off-by: Felix Nüsse <felix.nuesse@t-online.de>

newhinton committed:
change createCommand to toArray()

newhinton committed:
autoformat code in syncservice

aluo-x commented:
I just looked through the change log, this changes seem very extensive! However the changeset does seem to focus on tasks - is there a common way on Android to trigger on event (new file/photo?) without being killed by any power conservation process?

newhinton commented:
@aluo-x Good question! Generally it should not be hard to add additional triggers, like "on connection to wifi" or similar, but i am not sure if there is a good way to listen for changes on the disk. There is [a FileObserver](https://developer.android.com/reference/android/os/FileObserver) but it seems we would need to keep rcx permanently alive in the background. 

And either way, first this needs to get merged to master before i add such features.

newhinton committed:
update notification to use proper progressbar

Signed-off-by: Felix Nüsse <felix.nuesse@t-online.de>

newhinton committed:
add translations to new notification

Signed-off-by: Felix Nüsse <felix.nuesse@t-online.de>

AtilioA commented:
Any update on this? 👀

newhinton commented:
@AtilioA Sadly no, but if you want to test it anyway, take a look here: https://github.com/x0b/rcx/issues/38#issuecomment-1008675703

newhinton commented:
@x0b Heyho! Do you think you could give me a rough timeframe if/when you can review this PR and if you need changes beforehand? I'd like to make other changes, and i need to decide wether to wait for a review or if i start maintaining my own fork.

It's not that i cant wait further if it takes a bit more time, but it would be good to know if it is a week or a year, so that i know whats up with rcx!


Thank you very much! 


Edit: I know this is by no means a small PR. However, sometimes there are big changes which cant be easily split. Also, many changes are layouts or boilerplate, so the actual code is considerably less, and it is easily splittable for a review. I am also working on refactoring code even further so that it can be split even further while reviewing

newhinton committed:
Fix Taskname beeing empty on 'bad' Remote-Path

newhinton committed:
add retry button

Signed-off-by: Felix Nüsse <felix.nuesse@t-online.de>

newhinton committed:
unify internalTask start intent
move notifications to helper class

Signed-off-by: Felix Nüsse <felix.nuesse@t-online.de>

newhinton committed:
Fix TaskStartService not having proper notification

Signed-off-by: Felix Nüsse <felix.nuesse@t-online.de>

newhinton committed:
readd grouped sync notifications

Signed-off-by: Felix Nüsse <felix.nuesse@t-online.de>

newhinton committed:
clean up imports

Signed-off-by: Felix Nüsse <felix.nuesse@t-online.de>

newhinton committed:
use env

Signed-off-by: Felix Nüsse <felix.nuesse@t-online.de>

newhinton committed:
properly calculate sdk location

newhinton committed:
Merge remote-tracking branch 'x0b/master'

# Conflicts:
#	app/src/main/java/ca/pkay/rcloneexplorer/MainActivity.java
#	app/src/main/res/values-de/strings.xml
#	app/src/main/res/values/strings.xml
#	rclone/build.gradle

newhinton committed:
document building in readme

newhinton committed:
Merge remote-tracking branch 'origin/feature_rclone_use_json'

# Conflicts:
#	app/src/main/java/ca/pkay/rcloneexplorer/Rclone.java
#	app/src/main/java/ca/pkay/rcloneexplorer/Services/SyncService.java
#	app/src/main/res/values/strings.xml

newhinton committed:
merge json notification into master

feilongfl committed:
build: update go version

feilongfl committed:
rclone: update rclone to 1.58.1

feilongfl committed:
build: update gradle to 7.2.1

feilongfl committed:
build: update ndk version

newhinton committed:
Fix Darkmode

newhinton committed:
add night-colors

newhinton committed:
change strings

newhinton committed:
Fix theming on mainpage

newhinton committed:
fix appbar size and overlap

newhinton committed:
Overhaul Triggerview

newhinton committed:
fix missing colors

newhinton committed:
use brachiosaurus

newhinton committed:
fix color references in trigger_item

newhinton committed:
remove background color on task-list-item

newhinton committed:
update navigation drawer design

newhinton committed:
update icons for sync

newhinton committed:
update setting views

newhinton committed:
Fix theming not beeing applied properly

newhinton committed:
allow SyncService to create better notifications

newhinton committed:
use ThemeHelper for Darkmode in FilePicker

newhinton committed:
Split onHandleIntent to be lean and the sync process to be callable

newhinton committed:
Remove TaskStartService

newhinton committed:
use more two-tone icons

newhinton committed:
make filepicker leaner

newhinton committed:
Moved Activities to owm Package

newhinton committed:
Start implementing RemoteFolderPickerFragment

newhinton committed:
remove tint from icon

newhinton committed:
Fix Breadcrumb and title

newhinton committed:
update translations

newhinton committed:
set proper initial path in RemoteFolderPickerFragment

newhinton committed:
fix color attribute beeing used instead of @color

newhinton committed:
Move notifications to notification package
Update Upload&Download-Notifications

newhinton committed:
remove accidential upload bandwithlimit

newhinton committed:
Update iconography

LuckyTurtleDev commented:
Are their are any plans plans for a fdroid release of extRact?
rcx seems not been maintained anymore.

newhinton committed:
Update README.md

newhinton commented:
@Lukas1818 I am not yet decided on that. First i need to figure out the licensing (I'd like to use only a single license, but i am not sure if that's allowed or why there was a second license at all.)

If there is demand i can build and provide apk's via the release, but first i need to make some changes to the code.

Edit:
Also, in some parts my changes are really messy at the moment because i have moved a lot of stuff around. So a refactor would be due before that.

Edit 2: 
Also i do not know how much of a 'maintainer' i would be. I will keep it working, because i need it to, and i will probably implement some features i have seen here, but it may be possible that i will not spend more time than required for basic maintenance.

newhinton committed:
remove duplicated code

newhinton committed:
Merge remote-tracking branch 'origin/master'

newhinton committed:
readd name to sync notification

newhinton committed:
add new icons

newhinton committed:
Update strings for notification

newhinton committed:
Merge remote-tracking branch 'origin/feature_tasklog'

# Conflicts:
#	app/src/main/java/ca/pkay/rcloneexplorer/Activities/MainActivity.java
#	app/src/main/java/ca/pkay/rcloneexplorer/Rclone.java
#	app/src/main/java/ca/pkay/rcloneexplorer/Services/SyncService.java
#	app/src/main/res/values/strings.xml

newhinton committed:
write logs from sync-service, upload-service and download-service

newhinton committed:
update readme

newhinton committed:
Update README.md

newhinton committed:
update icons

newhinton committed:
update log style

newhinton committed:
update log and notification text

newhinton committed:
remember last used fragment

newhinton committed:
add missing import

newhinton committed:
add more missing imports

newhinton committed:
update readme todo

newhinton committed:
allow errors to show up in notification and logs

newhinton committed:
make notification for sync more useful

newhinton committed:
allow trigger to run when idle

newhinton committed:
expose "allow while idle" in ui

newhinton committed:
add missing german translation

newhinton committed:
switch applicationId

newhinton committed:
change output-apk name

newhinton committed:
fix wrong datatype in json export of appsettings

newhinton committed:
properly set theme&color in MainActivity

newhinton committed:
Merge remote-tracking branch 'feilongfl/master'

# Conflicts:
#	build.gradle

newhinton committed:
update security email

newhinton committed:
merge feilongfl:master to update rclone to 1.58

newhinton committed:
up version

newhinton committed:
update navigation drawer styling

newhinton committed:
fix color state list for light theme

newhinton committed:
fix issue with json

newhinton committed:
update version

newhinton committed:
use cards in listviews
and update about-cards

newhinton committed:
fix wrong maintainer listed

newhinton committed:
fix download notification having wrong channelid

newhinton committed:
update dialog style to material3

newhinton committed:
apply material design theme to all dialogs

newhinton committed:
update targetsdk to 32

newhinton committed:
update main toolbar to material style

newhinton committed:
fix some color issues

newhinton committed:
update Readme

newhinton committed:
remove old screenshots and update readme again

newhinton committed:
make all pending intents immutable for api 32

newhinton committed:
move threedot menu on cards

newhinton committed:
Update handling of metered and unmetered connections

djpohly committed:
Update sync progress notifications quietly

Every progress update on a sync operation currently causes the notification to play sound, vibrate, etc.

newhinton committed:
allow standalone rclone.conf imports

newhinton committed:
Merge pull request #13 from djpohly/patch-1

Update sync progress notifications quietly

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/x0b/rcx?pullRequest=84) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/x0b/rcx?pullRequest=84) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: newhinton<br/>:x: djpohly<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/x0b/rcx?pullRequest=84) it.</sub>

newhinton committed:
Merge remote-tracking branch 'origin/master'

newhinton committed:
increase padding for appbar

newhinton committed:
dont use special chars for export-filename

newhinton committed:
update filepicker for new design

newhinton committed:
fix nullpointer on no folder selected

newhinton committed:
Fix problematic remote-path behavior for task creation

newhinton committed:
add back button to task_activity

newhinton committed:
add new style to settings and trigger activity

newhinton committed:
make card radius adjustable

newhinton committed:
update task creation layout

newhinton committed:
fix the immutability flag

newhinton committed:
prepare bidirectional sync

newhinton committed:
implement descriptions for sync in code

newhinton committed:
dont start path with more than one backslash

newhinton committed:
fix title for TaskItem

newhinton committed:
Update Task Design again

newhinton committed:
overhaul intro

newhinton committed:
add deletions to log

newhinton committed:
disable bidirectional sync for now

newhinton committed:
Append Files that error out to notification and log

newhinton committed:
fix tint in fragment_permission_item.xml

newhinton committed:
fix storage access framework

newhinton committed:
update version

newhinton committed:
fix oauth not using correct io-stream

newhinton committed:
dont start arbitrary fragment when onStart is called to preserve fragments

newhinton committed:
dont resume the last fragment. Let android handle it.

newhinton committed:
send proper cancel intents for upload&download

newhinton committed:
reduce retries and timeouts

newhinton committed:
remove duplicate rclone parameter

newhinton committed:
migrate TaskActivity to Kotlin
fixed remotePath not keeping path
fixes #33

newhinton commented:
I have created a fork and development continues over at my repository.

(If i have any signed CLA's, please consider them revoked for now.)
